### PR TITLE
Make sure stack is aligned to 16 bytes

### PR DESCRIPTION
**Pull request by ggawryal**
We're using stack pointer modifying instructions (add directly to `RSP`, `StackPush`, `StackPop` and `Call`) only in `genPrologue, genEpilogue` and `genCall` functions. 
The stack will be alligned to 16 bytes at the entry of each program, so all we have to do is to:
1) move RSP by 8 bytes at thebeginning of `genCall` if some function has more than an odd, larger than 6 number of arguments
2) make sure allocated space for non-local and spilled variables in `genPrologue` is divisible by 16.

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate/Constant.kt [hunk 1]
```diff
@@ -15,3 +15,11 @@ class SummedConstant(private val first: Constant, private val second: Constant)
 
     override val value get() = first.value + second.value
 }
+
+class ConstantAlignedToAGivenRestModulo(private val rawValue: Constant, private val modulo: Long, private val rest: Long = 0L) : Constant {
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/intermediate/Constant.kt:19]: I think the word "remainder" is used instead of "rest" in this context.

**Review by tomekb234 — APPROVED** [review]

File: app/src/main/kotlin/compiler/intermediate/Constant.kt (hunk 1 above)
Comment (tomekb234) [reply app/src/main/kotlin/compiler/intermediate/Constant.kt:19]: I'd suggest to simply name this as `AlignedConstant`.

File: app/src/main/kotlin/compiler/intermediate/generators/SysV64CallingConvention.kt [hunk 2]
```diff
@@ -7,41 +7,59 @@ import compiler.intermediate.Register
...
+        val cfgBuilder = ControlFlowGraphBuilder()
+        // Align stack
+        if (stackMovement % stackAlignmentInBytes != 0L) {
+            val toAdd = stackAlignmentInBytes - stackMovement % stackAlignmentInBytes
+            cfgBuilder.addLinksFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IFTNode.Subtract(
+                    IFTNode.RegisterRead(Register.RSP),
+                    IFTNode.Const(toAdd)
+                )
```
Comment (bebidek) [inline app/src/main/kotlin/compiler/intermediate/generators/SysV64CallingConvention.kt:40]: You need to actually save the result to RSP (`IFTNode.RegisterWrite`).

File: app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt [hunk 3]
```diff
@@ -67,7 +68,7 @@ data class DefaultFunctionDetailsGenerator(
         cfgBuilder.addLinksFromAllFinalRoots(
             CFGLinkType.UNCONDITIONAL,
             IFTNode.StackPush(IFTNode.RegisterRead(Register.RBP))
-        )
+        ) // disaligns stack
```
Comment (bebidek) [inline app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt:71]: Actually, this PUSH aligns the stack to 16B. SysV64 convention says that right after CALL, RSP % 16 = 8, so the push makes it aligned.

Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt:71]: Consequently, I am not sure whether the  whole program is correct as:
At this point (after PUSH) stack is aligned.
After local variable allocation stack is 8 modulo 16.
After pushing the arguments (while calling some other function) the stack is still 8 modulo 16.
After CALL, stack is aligned.
After PUSH, stack is not aligned.

**Review by bebidek — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt (hunk 3 above)
Comment (tomekb234) [reply app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt:71]: Indeed, it seems that right now CALLs are at 8 mod 16. Changing the remainder to 0 at line 90 should fix the problem.

File: app/src/main/kotlin/compiler/intermediate/Constant.kt (hunk 1 above)
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate/Constant.kt]: You're right. Applied both suggestions and added a short doc comment

File: app/src/main/kotlin/compiler/intermediate/generators/SysV64CallingConvention.kt (hunk 2 above)
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate/generators/SysV64CallingConvention.kt]: Great catch, fixed

File: app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt (hunk 3 above)
Comment (ggawryal) [reply app/src/main/kotlin/compiler/intermediate/generators/DefaultFunctionDetailsGenerator.kt]: Thanks for spotting, fixed

**Review by bebidek — APPROVED** [review]

**Pull request merged**